### PR TITLE
Tweak Distributed API

### PR DESCRIPTION
**Pull request by glwagner**
This PR changes up the API for `Distributed`. The changes are:

- Removal of the kwarg `enable_overlapping_computation` which doesn't belong in architectures (which are model agnostic). (We can add that feature to specific models if desired.)
- Automatic initialization if `MPI.Initialized()` is false
- Implementation of default `ranks` (by default we distribute in `x`)

Perhaps more to come.

I'm also going to clean up up the examples, since they are the only source of information about running distributed simulations currently.

### CODE REVIEW

**Comment by glwagner** [comment]
Are there any tests for distributed hydrostatic models?

**Comment by simone-silvestri** [comment]
yeah in `test/test_distributed_models.jl`

**Comment by glwagner** [comment]
@simone-silvestri `DistributedComputations` is imported before models here:

https://github.com/CliMA/Oceananigans.jl/blob/82503b811a305bd7a6c4f1d3650f2b25b535c8fa/src/Oceananigans.jl#L221-L224

What was "enable overlapping communication" intended to support? Is there something that depends on whether `isnothing(mpi_requests)`? Perhaps more direct specification of this feature would help.

**Comment by glwagner** [comment]
> yeah in `test/test_distributed_models.jl`

There's no tests for `HydrostaticFreeSurfaceModel` there currently, just `NonhydrostaticModel`. 

Is there another file with tests for `HydrostaticFreeSurfaceModel`?

**Comment by glwagner** [comment]
How important is it to call `MPI.Finalize`? What happens if we don't do it? If it's critical, how do we ensure this happens? (This seems like a bad design...)

**Comment by glwagner** [comment]
I think users should put the global grid size into the `grid` constructor, not the local size. If the partitioning is irregular, this should go into the architecture. The reason we need this is to support a "one line" change from local to distributed computations. How do we implement this?

https://github.com/CliMA/Oceananigans.jl/blob/d7b97283b7b352b0c5c4c8ae4c7b56de65a9105c/src/DistributedComputations/distributed_grids.jl#L39

**Comment by glwagner** [comment]
Ok, I found in the docs for MPI.jl that

> Calling [MPI.Finalize()](https://juliaparallel.org/MPI.jl/stable/reference/environment/#MPI.Finalize) at the end of the program is optional, as it will be called automatically when Julia exits.

This is good news. So we can call `MPI.Init()` when we build the architecture, and avoid `MPI.Finalize()`. This brings us one step closer to the "one line to distributed simulations" goal.

https://github.com/JuliaParallel/MPI.jl

**Comment by glwagner** [comment]
Hopefully we can close #3318 here too, let's discuss.

File: validation/distributed_simulations/distributed_geostrophic_adjustment.jl [hunk 1]
```diff
@@ -16,76 +16,60 @@ using Oceananigans.Units: kilometers, meters
...
-
 topo = (Bounded, Periodic, Bounded)
-arch = Distributed(CPU(); topology = topo, 
-                 ranks=(Nranks, 1, 1),
-                 use_buffers = true)
+arch = Distributed(CPU(); topology = topo)
+
+# Distribute problem irregularly
+Nx = [10, 13, 18, 39]
+rank = MPI.Comm_rank(arch.communicator)
```
Comment (simone-silvestri) [inline validation/distributed_simulations/distributed_geostrophic_adjustment.jl:24]:
```suggestion

partition = Partition([10, 13, 18, 39])

arch = Distributed(CPU(); topology = topo, partition)

# Distribute problem irregularly
Nx = 80
rank = MPI.Comm_rank(arch.communicator)
```

File: validation/distributed_simulations/distributed_geostrophic_adjustment.jl [hunk 2]
```diff
@@ -16,76 +16,60 @@ using Oceananigans.Units: kilometers, meters
...
 
 Lh = 100kilometers
 Lz = 400meters
 
-Nx = [10, 13, 18, 39]
-
 grid = RectilinearGrid(arch,
                        size = (Nx[rank+1], 3, 1),
```
Comment (simone-silvestri) [inline validation/distributed_simulations/distributed_geostrophic_adjustment.jl:30]:
```suggestion
                       size = (Nx, 3, 1),
```

File: src/DistributedComputations/distributed_architectures.jl [hunk 3]
```diff
@@ -4,17 +4,87 @@ using CUDA: ndevices, device!
...
+
+Return `Partition` representing the division of a domain into
+`Rx` parts in `x` and `Ry` parts in `y` and `Rz` parts in `z`,
+where `x, y, z` are the first, second, and third dimension
+respectively.
+"""
+Partition(Rx::Number, Ry::Number=1, Rz::Number=1) = Partition{Rx, Ry, Rz}()
+Partition(lengths::Array{Int, 1}) = Partition(reshape(lengths, length(lengths), 1))
+Base.size(::Partition{<:Any, Rx, Ry, Rz}) where {Rx, Ry, Rz} = (Rx, Ry, Rz)
+
```
Comment (simone-silvestri) [inline src/DistributedComputations/distributed_architectures.jl:77]:
```suggestion

Partition(; Rx = 1, Ry = 1, Rz = 1) = Partition(Rx, Ry, Rz) 
```

File: src/DistributedComputations/distributed_architectures.jl [hunk 4]
```diff
@@ -58,50 +123,49 @@ Keyword arguments
                   if not for testing or developing. Change at your own risk!
 """
 function Distributed(child_architecture = CPU(); 
-                         topology, 
-                         ranks,
-                         devices = nothing, 
-                         enable_overlapped_computation = true,
```
Comment (simone-silvestri) [inline src/DistributedComputations/distributed_architectures.jl:64]: Now the nonhydrostatic and hydrostatic model will always be working in asynchronous mode?

**Comment by simone-silvestri** [comment]
https://github.com/CliMA/Oceananigans.jl/blob/5dd96cf310e988fc5d3a2523c2dcc7197f18ff62/src/DistributedComputations/distributed_grids.jl#L326 

here we probably need something like
```
function scatter_local_grids(arch::Distributed, global_grid::RectilinearGrid, local_size)
    x, y, z, topo, halo = scatter_grid_properties(global_grid)
    global_size = map(sum, concatenate_local_sizes(local_size, arch))
    return RectilinearGrid(arch, eltype(global_grid); size=global_size, x=x, y=y, z=z, halo=halo, topology=topo)
end
```

same for the `LatitudeLongitudeGrid`

File: src/DistributedComputations/distributed_architectures.jl (hunk 4 above)
Comment (glwagner) [reply src/DistributedComputations/distributed_architectures.jl:64]: Neither is really “supported”, there’s no code that tests the difference between the two anyways. But in a future PR we can add an option in those model constructors if we think users will want to toggle between the two.

**Comment by glwagner** [comment]
I don’t think we should fix the untested features in this PR. We need to merge this ASAP so we can open a new PR with tests.

**Comment by glwagner** [comment]
@simone-silvestri can you fix the failing test? As soon as tests pass please merge. We need a another PR that implements tests following this one.

File: src/DistributedComputations/distributed_architectures.jl [hunk 5]
```diff
@@ -4,17 +4,106 @@ using CUDA: ndevices, device!
...
+                   mpi_tag :: T) where {S, A, Δ, R, ρ, I, C, γ, M, T} = 
+                   new{A, S, Δ, R, ρ, I, C, γ, M, T}(child_architecture,
+                                                     partition,
+                                                     ranks,
+                                                     local_rank,
+                                                     local_index,
+                                                     connectivity,
+                                                     communicator,
+                                                     mpi_requests,
+                                                     mpi_tag)
```
Comment (navidcy) [inline src/DistributedComputations/distributed_architectures.jl:106]:
```suggestion
                       new{A, S, Δ, R, ρ, I, C, γ, M, T}(child_architecture,
                                                         partition,
                                                         ranks,
                                                         local_rank,
                                                         local_index,
                                                         connectivity,
                                                         communicator,
                                                         mpi_requests,
                                                         mpi_tag)
```

File: src/DistributedComputations/distributed_fields.jl [hunk 6]
```diff
@@ -31,4 +31,28 @@ function set!(u::DistributedField, f::Function)
...
+        arch.mpi_tag[] -= arch.mpi_tag[]
+    
+        # Reset MPI requests
+        empty!(arch.mpi_requests)
+    end
+    
+    recv_from_buffers!(field.data, field.boundary_buffers, field.grid)
+    
+    return nothing
 end
```
Comment (navidcy) [inline src/DistributedComputations/distributed_fields.jl:58]:
```suggestion
    return nothing
end

```

File: src/DistributedComputations/distributed_grids.jl [hunk 7]
```diff
@@ -84,7 +99,7 @@ function LatitudeLongitudeGrid(arch::Distributed,
                                radius = R_Earth,
                                halo = (1, 1, 1))
 
-    global_size = map(sum, concatenate_local_sizes(size, arch))
+    # global_size = map(sum, concatenate_local_sizes(size, arch))
```
Comment (navidcy) [inline src/DistributedComputations/distributed_grids.jl:102]: why comment out? delete?

File: src/ImmersedBoundaries/distributed_immersed_boundaries.jl [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        cpu_bottom   = arch_array(CPU(), ib.bottom_height)[1:Nx, 1:Ny] 
+        set!(bottom_field, cpu_bottom)
+        fill_halo_regions!(bottom_field)
+        offset_bottom_array = dropdims(bottom_field.data, dims=3)
+
+        return getnamewrapper(ib)(offset_bottom_array)
+    end
+    
+    return ib
+end
```
Comment (navidcy) [inline src/ImmersedBoundaries/distributed_immersed_boundaries.jl:87]:
```suggestion
end

```

File: src/Oceananigans.jl [hunk 9]
```diff
@@ -218,7 +222,7 @@ include("Biogeochemistry.jl")
 
 # TODO: move above
 include("ImmersedBoundaries/ImmersedBoundaries.jl")
-include("DistributedComputations/DistributedComputations.jl")
+# include("DistributedComputations/DistributedComputations.jl")
```
Comment (navidcy) [inline src/Oceananigans.jl:225]: why comment out?

File: src/DistributedComputations/distributed_fields.jl (hunk 6 above)
Comment (navidcy) [inline src/DistributedComputations/distributed_fields.jl:58]:
```suggestion
end

```

File: test/test_distributed_models.jl [hunk 10]
```diff
@@ -483,8 +485,9 @@ end
     @testset "Time stepping ShallowWaterModel" begin
         for child_arch in archs
             topo = (Periodic, Periodic, Flat)
-            arch = Distributed(child_arch; ranks=(1, 4, 1), topology = topo, devices = (0, 0, 0, 0))
-            grid = RectilinearGrid(arch, topology=topo, size=(8, 2), extent=(1, 2), halo=(3, 3))
+            #arch = Distributed(child_arch; ranks=(1, 4, 1), topology = topo, devices = (0, 0, 0, 0))
```
Comment (navidcy) [inline test/test_distributed_models.jl:488]: delete the comment?

File: validation/distributed_simulations/distributed_shallow_water_turbulence.jl [hunk 11]
```diff
@@ -1,61 +1,55 @@
...
 uh, vh, h = model.solution
-outputs = (ζ=Field(∂x(vh/h) - ∂y(uh/h)),)
+outputs = (; ζ=Field(∂x(vh/h) - ∂y(uh/h)))
 filepath = "mpi_shallow_water_turbulence_rank$(local_rank).nc"
-simulation.output_writers[:fields] =
-    NetCDFOutputWriter(model, outputs, filepath=filepath, schedule=TimeInterval(1.0), mode="c")
+simulation.output_writers[:fields] = NetCDFOutputWriter(model, outputs, filepath=filepath, schedule=TimeInterval(1), mode="c")
 
-MPI.Barrier(MPI.COMM_WORLD)
+# ?
```
Comment (navidcy) [inline validation/distributed_simulations/distributed_shallow_water_turbulence.jl:39]: `# ?` ?

**Review by simone-silvestri — APPROVED** [review]

**Pull request merged**